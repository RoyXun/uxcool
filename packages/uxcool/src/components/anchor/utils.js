const REG_ANCHOR_ID = /#[^#]+$/;

export function getAnchorRelateElement(href = '') {
  const matches = href.match(REG_ANCHOR_ID);
  if (!matches) {
    return null;
  }

  return document.querySelector(matches[0]);
}

export function getElementOffsetTop(element, container) {
  let top = 0;
  if (!element || !container) {
    return top;
  }

  const box = element.getBoundingClientRect();
  if (container.window === window) {
    const doc = element.ownerDocument;
    const docElement = doc && doc.documentElement;
    const body = doc && doc.body;
    top = box.top - docElement.clientTop || body.clientTop || 0;
  } else {
    const containerRect = container.getBoundingClientRect();
    top = box.top - containerRect.top;
  }
  return top;
}
