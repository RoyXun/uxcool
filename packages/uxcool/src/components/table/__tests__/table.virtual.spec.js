import { mountPickerFactory, triggerEvent, getVueCreateElement } from '@suning/v-test-utils';
import { getScrollCols, getScrollData } from './utils';
import Table from '../table';

describe('Table.virtual', () => {
  jest.setTimeout(100000);
  const mountPicker = mountPickerFactory(Table);
  const columns = getScrollCols(5);
  const dataSource = getScrollData(200);
  it('render virtual table ', async () => {
    const wrapper = await mountPicker({
      propsData: {
        value: dataSource,
        columns,
        virtualScroll: {
          itemSize: 40,
        },
      },
    });
    expect(wrapper.html()).toMatchSnapshot();
  });

  describe('scroll', () => {
    it('scroll.x is true', async () => {
      const wrapper = await mountPicker({
        propsData: {
          value: dataSource,
          columns,
          scroll: {
            x: true,
          },
          virtualScroll: {
            itemSize: 40,
          },
        },
      });

      expect(wrapper.html()).toMatchSnapshot();
    });

    it('scroll.x is number', async () => {
      const wrapper = await mountPicker({
        propsData: {
          value: dataSource,
          columns,
          scroll: {
            x: 200,
          },
          virtualScroll: {
            itemSize: 40,
          },
        },
      });

      expect(wrapper.html()).toMatchSnapshot();
    });

    it('scroll.y is number', async () => {
      const wrapper = await mountPicker({
        propsData: {
          value: dataSource,
          columns,
          scroll: {
            y: 200,
          },
          virtualScroll: {
            itemSize: 40,
          },
        },
      });

      expect(wrapper.html()).toMatchSnapshot();
    });

    it('itemSize = 40', async () => {
      const wrapper = await mountPicker({
        propsData: {
          value: dataSource,
          columns,
          virtualScroll: {
            itemSize: 40,
          },
        },
      });

      const bTable = wrapper.find('.ux-table-main-table').find('.ux-table-body');

      bTable.element.scrollTop = 1000;

      await triggerEvent(bTable, 'scroll', 120);

      expect(wrapper.html()).toMatchSnapshot();
    });

    it('itemSize is auto', async () => {
      const wrapper = await mountPicker({
        propsData: {
          value: dataSource,
          columns,
          virtualScroll: true,
        },
      });

      const bTable = wrapper.find('.ux-table-main-table').find('.ux-table-body');

      bTable.element.scrollTop = 1000;

      await triggerEvent(bTable, 'scroll', 120);
      expect(parseInt(wrapper.find('[role=virtual]').element.style.paddingTop, 10)).toBeGreaterThan(
        0
      );
    });

    it('scroll and fixed column', async () => {
      const cols = getScrollCols();
      cols[0].fixed = 'left';
      cols[cols.length - 1].fixed = 'right';
      const wrapper = await mountPicker({
        propsData: {
          value: dataSource,
          columns: cols,
          scroll: {
            x: 200,
            y: 200,
          },
          virtualScroll: {
            itemSize: 40,
          },
        },
      });

      const bTable = wrapper.find('.ux-table-main-table').find('.ux-table-body');

      bTable.element.scrollTop = 1000;

      await triggerEvent(bTable, 'scroll', 120);
      expect(parseInt(wrapper.find('[role=virtual]').element.style.paddingTop, 10)).toBeGreaterThan(
        0
      );
      expect(wrapper.find('.ux-table-fixed-left').exists()).toBeTruthy();
      expect(wrapper.find('.ux-table-fixed-right').exists()).toBeTruthy();
    });
  });

  describe('expand', () => {
    const mockText = 'expand render test text';
    const h = getVueCreateElement();
    it('expanded-row-render', async () => {
      const onExpand = jest.fn();
      const wrapper = await mountPicker({
        propsData: {
          value: dataSource,
          columns,
          virtualScroll: true,
          expandedRowRender() {
            return h('p', { class: 'expanded-render-test' }, [mockText]);
          },
        },
        listeners: {
          expand: onExpand,
        },
      });
      await triggerEvent(wrapper.find('.ux-table-row-expand-icon'), 'click');

      expect(onExpand).toHaveBeenCalled();
      expect(onExpand.mock.calls[0][0]).toBeTruthy();
      expect(wrapper.find('.expanded-render-test').exists()).toBeTruthy();
    });
  });

  it('fixed table hover', async () => {
    const cols = getScrollCols();
    cols[0].fixed = 'left';
    cols[cols.length - 1].fixed = 'right';

    const wrapper = await mountPicker({
      propsData: {
        value: dataSource,
        columns: cols,
        virtualScroll: true,
      },
    });

    const mainTables = wrapper.findAll('.ux-table-main-table');

    const mainTableRows = mainTables.at(0).findAll('.ux-table-tbody tr');
    const leftTableRows = mainTables.at(1).findAll('.ux-table-tbody tr');
    const rightTableRows = mainTables.at(2).findAll('.ux-table-tbody tr');

    await triggerEvent(mainTableRows.at(0), 'mouseenter');

    expect(leftTableRows.at(0).classes('ux-table-row-hover')).toBeTruthy();
    expect(rightTableRows.at(0).classes('ux-table-row-hover')).toBeTruthy();

    await triggerEvent(mainTableRows.at(0), 'mouseleave');
    expect(leftTableRows.at(0).classes('ux-table-row-hover')).toBeFalsy();
    expect(rightTableRows.at(0).classes('ux-table-row-hover')).toBeFalsy();

    await triggerEvent(leftTableRows.at(1), 'mouseenter');
    expect(mainTableRows.at(1).classes('ux-table-row-hover')).toBeTruthy();
    expect(rightTableRows.at(1).classes('ux-table-row-hover')).toBeTruthy();

    await triggerEvent(leftTableRows.at(1), 'mouseleave');
    expect(mainTableRows.at(1).classes('ux-table-row-hover')).toBeFalsy();
    expect(rightTableRows.at(1).classes('ux-table-row-hover')).toBeFalsy();

    await triggerEvent(rightTableRows.at(2), 'mouseenter');
    expect(mainTableRows.at(2).classes('ux-table-row-hover')).toBeTruthy();
    expect(leftTableRows.at(2).classes('ux-table-row-hover')).toBeTruthy();

    await triggerEvent(rightTableRows.at(2), 'mouseleave');
    expect(mainTableRows.at(2).classes('ux-table-row-hover')).toBeFalsy();
    expect(leftTableRows.at(2).classes('ux-table-row-hover')).toBeFalsy();
  });
});
