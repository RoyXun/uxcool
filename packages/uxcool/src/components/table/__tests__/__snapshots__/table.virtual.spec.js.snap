// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Table.virtual render virtual table  1`] = `
<div class="ux-table-wrapper ux-table-light">
  <div class="ux-spin-nested-loading">
    <div class="ux-spin" style="display: none;"><svg viewBox="25 25 50 50" class="ux-spin-circle">
        <circle cx="50" cy="50" r="20" fill="none"></circle>
      </svg> </div>
    <div class="ux-spin-container">
      <div class="ux-table ux-table-fixed-header ux-table-scroll-position-left ux-table-default">
        <div class="ux-table-content">
          <div class="ux-table-main-table ux-table-scroll">
            <div class="ux-table-header">
              <table class="">
                <colgroup>
                  <col>
                  <col>
                  <col>
                  <col>
                  <col>
                </colgroup>
                <thead class="ux-table-thead">
                  <tr>
                    <th>title-0</th>
                    <th>title-1</th>
                    <th>title-2</th>
                    <th>title-3</th>
                    <th>title-4</th>
                  </tr>
                </thead>
                <!---->
              </table>
            </div>
            <div class="ux-table-body" style="max-height: 500px; overflow-y: scroll;">
              <div role="virtual" style="padding-top: 0px; padding-bottom: 7280px;">
                <table class="">
                  <colgroup>
                    <col>
                    <col>
                    <col>
                    <col>
                    <col>
                  </colgroup>
                  <tbody class="ux-table-tbody">
                    <tr class="ux-table-row">
                      <td>k1111-0</td>
                      <td>k2222-0</td>
                      <td>k1111-0</td>
                      <td>k2222-0</td>
                      <td>k1111-0</td>
                    </tr>
                    <tr class="ux-table-row">
                      <td>k1111-1</td>
                      <td>k2222-1</td>
                      <td>k1111-1</td>
                      <td>k2222-1</td>
                      <td>k1111-1</td>
                    </tr>
                    <tr class="ux-table-row">
                      <td>k1111-2</td>
                      <td>k2222-2</td>
                      <td>k1111-2</td>
                      <td>k2222-2</td>
                      <td>k1111-2</td>
                    </tr>
                    <tr class="ux-table-row">
                      <td>k1111-3</td>
                      <td>k2222-3</td>
                      <td>k1111-3</td>
                      <td>k2222-3</td>
                      <td>k1111-3</td>
                    </tr>
                    <tr class="ux-table-row">
                      <td>k1111-4</td>
                      <td>k2222-4</td>
                      <td>k1111-4</td>
                      <td>k2222-4</td>
                      <td>k1111-4</td>
                    </tr>
                    <tr class="ux-table-row">
                      <td>k1111-5</td>
                      <td>k2222-5</td>
                      <td>k1111-5</td>
                      <td>k2222-5</td>
                      <td>k1111-5</td>
                    </tr>
                    <tr class="ux-table-row">
                      <td>k1111-6</td>
                      <td>k2222-6</td>
                      <td>k1111-6</td>
                      <td>k2222-6</td>
                      <td>k1111-6</td>
                    </tr>
                    <tr class="ux-table-row">
                      <td>k1111-7</td>
                      <td>k2222-7</td>
                      <td>k1111-7</td>
                      <td>k2222-7</td>
                      <td>k1111-7</td>
                    </tr>
                    <tr class="ux-table-row">
                      <td>k1111-8</td>
                      <td>k2222-8</td>
                      <td>k1111-8</td>
                      <td>k2222-8</td>
                      <td>k1111-8</td>
                    </tr>
                    <tr class="ux-table-row">
                      <td>k1111-9</td>
                      <td>k2222-9</td>
                      <td>k1111-9</td>
                      <td>k2222-9</td>
                      <td>k1111-9</td>
                    </tr>
                    <tr class="ux-table-row">
                      <td>k1111-10</td>
                      <td>k2222-10</td>
                      <td>k1111-10</td>
                      <td>k2222-10</td>
                      <td>k1111-10</td>
                    </tr>
                    <tr class="ux-table-row">
                      <td>k1111-11</td>
                      <td>k2222-11</td>
                      <td>k1111-11</td>
                      <td>k2222-11</td>
                      <td>k1111-11</td>
                    </tr>
                    <tr class="ux-table-row">
                      <td>k1111-12</td>
                      <td>k2222-12</td>
                      <td>k1111-12</td>
                      <td>k2222-12</td>
                      <td>k1111-12</td>
                    </tr>
                    <tr class="ux-table-row">
                      <td>k1111-13</td>
                      <td>k2222-13</td>
                      <td>k1111-13</td>
                      <td>k2222-13</td>
                      <td>k1111-13</td>
                    </tr>
                    <tr class="ux-table-row">
                      <td>k1111-14</td>
                      <td>k2222-14</td>
                      <td>k1111-14</td>
                      <td>k2222-14</td>
                      <td>k1111-14</td>
                    </tr>
                    <tr class="ux-table-row">
                      <td>k1111-15</td>
                      <td>k2222-15</td>
                      <td>k1111-15</td>
                      <td>k2222-15</td>
                      <td>k1111-15</td>
                    </tr>
                    <tr class="ux-table-row">
                      <td>k1111-16</td>
                      <td>k2222-16</td>
                      <td>k1111-16</td>
                      <td>k2222-16</td>
                      <td>k1111-16</td>
                    </tr>
                    <tr class="ux-table-row">
                      <td>k1111-17</td>
                      <td>k2222-17</td>
                      <td>k1111-17</td>
                      <td>k2222-17</td>
                      <td>k1111-17</td>
                    </tr>
                  </tbody>
                </table>
              </div>
            </div>
            <!---->
            <!---->
          </div>
        </div>
        <div class="ux-table-resize-line" style="display: none;"></div>
      </div>
    </div>
  </div>
</div>
`;

exports[`Table.virtual scroll itemSize = 40 1`] = `
<div class="ux-table-wrapper ux-table-light">
  <div class="ux-spin-nested-loading">
    <div class="ux-spin" style="display: none;"><svg viewBox="25 25 50 50" class="ux-spin-circle">
        <circle cx="50" cy="50" r="20" fill="none"></circle>
      </svg> </div>
    <div class="ux-spin-container">
      <div class="ux-table ux-table-fixed-header ux-table-scroll-position-left ux-table-default">
        <div class="ux-table-content">
          <div class="ux-table-main-table ux-table-scroll">
            <div class="ux-table-header">
              <table class="">
                <colgroup>
                  <col>
                  <col>
                  <col>
                  <col>
                  <col>
                </colgroup>
                <thead class="ux-table-thead">
                  <tr>
                    <th>title-0</th>
                    <th>title-1</th>
                    <th>title-2</th>
                    <th>title-3</th>
                    <th>title-4</th>
                  </tr>
                </thead>
                <!---->
              </table>
            </div>
            <div class="ux-table-body" style="max-height: 500px; overflow-y: scroll;">
              <div role="virtual" style="padding-top: 800px; padding-bottom: 6280px;">
                <table class="">
                  <colgroup>
                    <col>
                    <col>
                    <col>
                    <col>
                    <col>
                  </colgroup>
                  <tbody class="ux-table-tbody">
                    <tr class="ux-table-row">
                      <td>k1111-20</td>
                      <td>k2222-20</td>
                      <td>k1111-20</td>
                      <td>k2222-20</td>
                      <td>k1111-20</td>
                    </tr>
                    <tr class="ux-table-row">
                      <td>k1111-21</td>
                      <td>k2222-21</td>
                      <td>k1111-21</td>
                      <td>k2222-21</td>
                      <td>k1111-21</td>
                    </tr>
                    <tr class="ux-table-row">
                      <td>k1111-22</td>
                      <td>k2222-22</td>
                      <td>k1111-22</td>
                      <td>k2222-22</td>
                      <td>k1111-22</td>
                    </tr>
                    <tr class="ux-table-row">
                      <td>k1111-23</td>
                      <td>k2222-23</td>
                      <td>k1111-23</td>
                      <td>k2222-23</td>
                      <td>k1111-23</td>
                    </tr>
                    <tr class="ux-table-row">
                      <td>k1111-24</td>
                      <td>k2222-24</td>
                      <td>k1111-24</td>
                      <td>k2222-24</td>
                      <td>k1111-24</td>
                    </tr>
                    <tr class="ux-table-row">
                      <td>k1111-25</td>
                      <td>k2222-25</td>
                      <td>k1111-25</td>
                      <td>k2222-25</td>
                      <td>k1111-25</td>
                    </tr>
                    <tr class="ux-table-row">
                      <td>k1111-26</td>
                      <td>k2222-26</td>
                      <td>k1111-26</td>
                      <td>k2222-26</td>
                      <td>k1111-26</td>
                    </tr>
                    <tr class="ux-table-row">
                      <td>k1111-27</td>
                      <td>k2222-27</td>
                      <td>k1111-27</td>
                      <td>k2222-27</td>
                      <td>k1111-27</td>
                    </tr>
                    <tr class="ux-table-row">
                      <td>k1111-28</td>
                      <td>k2222-28</td>
                      <td>k1111-28</td>
                      <td>k2222-28</td>
                      <td>k1111-28</td>
                    </tr>
                    <tr class="ux-table-row">
                      <td>k1111-29</td>
                      <td>k2222-29</td>
                      <td>k1111-29</td>
                      <td>k2222-29</td>
                      <td>k1111-29</td>
                    </tr>
                    <tr class="ux-table-row">
                      <td>k1111-30</td>
                      <td>k2222-30</td>
                      <td>k1111-30</td>
                      <td>k2222-30</td>
                      <td>k1111-30</td>
                    </tr>
                    <tr class="ux-table-row">
                      <td>k1111-31</td>
                      <td>k2222-31</td>
                      <td>k1111-31</td>
                      <td>k2222-31</td>
                      <td>k1111-31</td>
                    </tr>
                    <tr class="ux-table-row">
                      <td>k1111-32</td>
                      <td>k2222-32</td>
                      <td>k1111-32</td>
                      <td>k2222-32</td>
                      <td>k1111-32</td>
                    </tr>
                    <tr class="ux-table-row">
                      <td>k1111-33</td>
                      <td>k2222-33</td>
                      <td>k1111-33</td>
                      <td>k2222-33</td>
                      <td>k1111-33</td>
                    </tr>
                    <tr class="ux-table-row">
                      <td>k1111-34</td>
                      <td>k2222-34</td>
                      <td>k1111-34</td>
                      <td>k2222-34</td>
                      <td>k1111-34</td>
                    </tr>
                    <tr class="ux-table-row">
                      <td>k1111-35</td>
                      <td>k2222-35</td>
                      <td>k1111-35</td>
                      <td>k2222-35</td>
                      <td>k1111-35</td>
                    </tr>
                    <tr class="ux-table-row">
                      <td>k1111-36</td>
                      <td>k2222-36</td>
                      <td>k1111-36</td>
                      <td>k2222-36</td>
                      <td>k1111-36</td>
                    </tr>
                    <tr class="ux-table-row">
                      <td>k1111-37</td>
                      <td>k2222-37</td>
                      <td>k1111-37</td>
                      <td>k2222-37</td>
                      <td>k1111-37</td>
                    </tr>
                    <tr class="ux-table-row">
                      <td>k1111-38</td>
                      <td>k2222-38</td>
                      <td>k1111-38</td>
                      <td>k2222-38</td>
                      <td>k1111-38</td>
                    </tr>
                    <tr class="ux-table-row">
                      <td>k1111-39</td>
                      <td>k2222-39</td>
                      <td>k1111-39</td>
                      <td>k2222-39</td>
                      <td>k1111-39</td>
                    </tr>
                    <tr class="ux-table-row">
                      <td>k1111-40</td>
                      <td>k2222-40</td>
                      <td>k1111-40</td>
                      <td>k2222-40</td>
                      <td>k1111-40</td>
                    </tr>
                    <tr class="ux-table-row">
                      <td>k1111-41</td>
                      <td>k2222-41</td>
                      <td>k1111-41</td>
                      <td>k2222-41</td>
                      <td>k1111-41</td>
                    </tr>
                    <tr class="ux-table-row">
                      <td>k1111-42</td>
                      <td>k2222-42</td>
                      <td>k1111-42</td>
                      <td>k2222-42</td>
                      <td>k1111-42</td>
                    </tr>
                  </tbody>
                </table>
              </div>
            </div>
            <!---->
            <!---->
          </div>
        </div>
        <div class="ux-table-resize-line" style="display: none;"></div>
      </div>
    </div>
  </div>
</div>
`;

exports[`Table.virtual scroll scroll.x is number 1`] = `
<div class="ux-table-wrapper ux-table-light">
  <div class="ux-spin-nested-loading">
    <div class="ux-spin" style="display: none;"><svg viewBox="25 25 50 50" class="ux-spin-circle">
        <circle cx="50" cy="50" r="20" fill="none"></circle>
      </svg> </div>
    <div class="ux-spin-container">
      <div class="ux-table ux-table-fixed-header ux-table-scroll-position-left ux-table-default">
        <div class="ux-table-content">
          <div class="ux-table-main-table ux-table-scroll">
            <div class="ux-table-header">
              <table class="ux-table-fixed" style="table-layout: fixed; width: 200px;">
                <colgroup>
                  <col>
                  <col>
                  <col>
                  <col>
                  <col>
                </colgroup>
                <thead class="ux-table-thead">
                  <tr>
                    <th>title-0</th>
                    <th>title-1</th>
                    <th>title-2</th>
                    <th>title-3</th>
                    <th>title-4</th>
                  </tr>
                </thead>
                <!---->
              </table>
            </div>
            <div class="ux-table-body" style="overflow-x: auto; max-height: 500px; overflow-y: scroll; webkit-transform: translate3d(0,0,0);">
              <div role="virtual" style="padding-top: 0px; padding-bottom: 7280px;">
                <table class="ux-table-fixed" style="table-layout: fixed; width: 200px;">
                  <colgroup>
                    <col>
                    <col>
                    <col>
                    <col>
                    <col>
                  </colgroup>
                  <tbody class="ux-table-tbody">
                    <tr class="ux-table-row">
                      <td>k1111-0</td>
                      <td>k2222-0</td>
                      <td>k1111-0</td>
                      <td>k2222-0</td>
                      <td>k1111-0</td>
                    </tr>
                    <tr class="ux-table-row">
                      <td>k1111-1</td>
                      <td>k2222-1</td>
                      <td>k1111-1</td>
                      <td>k2222-1</td>
                      <td>k1111-1</td>
                    </tr>
                    <tr class="ux-table-row">
                      <td>k1111-2</td>
                      <td>k2222-2</td>
                      <td>k1111-2</td>
                      <td>k2222-2</td>
                      <td>k1111-2</td>
                    </tr>
                    <tr class="ux-table-row">
                      <td>k1111-3</td>
                      <td>k2222-3</td>
                      <td>k1111-3</td>
                      <td>k2222-3</td>
                      <td>k1111-3</td>
                    </tr>
                    <tr class="ux-table-row">
                      <td>k1111-4</td>
                      <td>k2222-4</td>
                      <td>k1111-4</td>
                      <td>k2222-4</td>
                      <td>k1111-4</td>
                    </tr>
                    <tr class="ux-table-row">
                      <td>k1111-5</td>
                      <td>k2222-5</td>
                      <td>k1111-5</td>
                      <td>k2222-5</td>
                      <td>k1111-5</td>
                    </tr>
                    <tr class="ux-table-row">
                      <td>k1111-6</td>
                      <td>k2222-6</td>
                      <td>k1111-6</td>
                      <td>k2222-6</td>
                      <td>k1111-6</td>
                    </tr>
                    <tr class="ux-table-row">
                      <td>k1111-7</td>
                      <td>k2222-7</td>
                      <td>k1111-7</td>
                      <td>k2222-7</td>
                      <td>k1111-7</td>
                    </tr>
                    <tr class="ux-table-row">
                      <td>k1111-8</td>
                      <td>k2222-8</td>
                      <td>k1111-8</td>
                      <td>k2222-8</td>
                      <td>k1111-8</td>
                    </tr>
                    <tr class="ux-table-row">
                      <td>k1111-9</td>
                      <td>k2222-9</td>
                      <td>k1111-9</td>
                      <td>k2222-9</td>
                      <td>k1111-9</td>
                    </tr>
                    <tr class="ux-table-row">
                      <td>k1111-10</td>
                      <td>k2222-10</td>
                      <td>k1111-10</td>
                      <td>k2222-10</td>
                      <td>k1111-10</td>
                    </tr>
                    <tr class="ux-table-row">
                      <td>k1111-11</td>
                      <td>k2222-11</td>
                      <td>k1111-11</td>
                      <td>k2222-11</td>
                      <td>k1111-11</td>
                    </tr>
                    <tr class="ux-table-row">
                      <td>k1111-12</td>
                      <td>k2222-12</td>
                      <td>k1111-12</td>
                      <td>k2222-12</td>
                      <td>k1111-12</td>
                    </tr>
                    <tr class="ux-table-row">
                      <td>k1111-13</td>
                      <td>k2222-13</td>
                      <td>k1111-13</td>
                      <td>k2222-13</td>
                      <td>k1111-13</td>
                    </tr>
                    <tr class="ux-table-row">
                      <td>k1111-14</td>
                      <td>k2222-14</td>
                      <td>k1111-14</td>
                      <td>k2222-14</td>
                      <td>k1111-14</td>
                    </tr>
                    <tr class="ux-table-row">
                      <td>k1111-15</td>
                      <td>k2222-15</td>
                      <td>k1111-15</td>
                      <td>k2222-15</td>
                      <td>k1111-15</td>
                    </tr>
                    <tr class="ux-table-row">
                      <td>k1111-16</td>
                      <td>k2222-16</td>
                      <td>k1111-16</td>
                      <td>k2222-16</td>
                      <td>k1111-16</td>
                    </tr>
                    <tr class="ux-table-row">
                      <td>k1111-17</td>
                      <td>k2222-17</td>
                      <td>k1111-17</td>
                      <td>k2222-17</td>
                      <td>k1111-17</td>
                    </tr>
                  </tbody>
                </table>
              </div>
            </div>
            <!---->
            <!---->
          </div>
        </div>
        <div class="ux-table-resize-line" style="display: none;"></div>
      </div>
    </div>
  </div>
</div>
`;

exports[`Table.virtual scroll scroll.x is true 1`] = `
<div class="ux-table-wrapper ux-table-light">
  <div class="ux-spin-nested-loading">
    <div class="ux-spin" style="display: none;"><svg viewBox="25 25 50 50" class="ux-spin-circle">
        <circle cx="50" cy="50" r="20" fill="none"></circle>
      </svg> </div>
    <div class="ux-spin-container">
      <div class="ux-table ux-table-fixed-header ux-table-scroll-position-left ux-table-default">
        <div class="ux-table-content">
          <div class="ux-table-main-table ux-table-scroll">
            <div class="ux-table-header">
              <table class="ux-table-fixed" style="table-layout: fixed; width: 100%;">
                <colgroup>
                  <col>
                  <col>
                  <col>
                  <col>
                  <col>
                </colgroup>
                <thead class="ux-table-thead">
                  <tr>
                    <th>title-0</th>
                    <th>title-1</th>
                    <th>title-2</th>
                    <th>title-3</th>
                    <th>title-4</th>
                  </tr>
                </thead>
                <!---->
              </table>
            </div>
            <div class="ux-table-body" style="overflow-x: auto; max-height: 500px; overflow-y: scroll; webkit-transform: translate3d(0,0,0);">
              <div role="virtual" style="padding-top: 0px; padding-bottom: 7280px;">
                <table class="ux-table-fixed" style="table-layout: fixed; width: 100%;">
                  <colgroup>
                    <col>
                    <col>
                    <col>
                    <col>
                    <col>
                  </colgroup>
                  <tbody class="ux-table-tbody">
                    <tr class="ux-table-row">
                      <td>k1111-0</td>
                      <td>k2222-0</td>
                      <td>k1111-0</td>
                      <td>k2222-0</td>
                      <td>k1111-0</td>
                    </tr>
                    <tr class="ux-table-row">
                      <td>k1111-1</td>
                      <td>k2222-1</td>
                      <td>k1111-1</td>
                      <td>k2222-1</td>
                      <td>k1111-1</td>
                    </tr>
                    <tr class="ux-table-row">
                      <td>k1111-2</td>
                      <td>k2222-2</td>
                      <td>k1111-2</td>
                      <td>k2222-2</td>
                      <td>k1111-2</td>
                    </tr>
                    <tr class="ux-table-row">
                      <td>k1111-3</td>
                      <td>k2222-3</td>
                      <td>k1111-3</td>
                      <td>k2222-3</td>
                      <td>k1111-3</td>
                    </tr>
                    <tr class="ux-table-row">
                      <td>k1111-4</td>
                      <td>k2222-4</td>
                      <td>k1111-4</td>
                      <td>k2222-4</td>
                      <td>k1111-4</td>
                    </tr>
                    <tr class="ux-table-row">
                      <td>k1111-5</td>
                      <td>k2222-5</td>
                      <td>k1111-5</td>
                      <td>k2222-5</td>
                      <td>k1111-5</td>
                    </tr>
                    <tr class="ux-table-row">
                      <td>k1111-6</td>
                      <td>k2222-6</td>
                      <td>k1111-6</td>
                      <td>k2222-6</td>
                      <td>k1111-6</td>
                    </tr>
                    <tr class="ux-table-row">
                      <td>k1111-7</td>
                      <td>k2222-7</td>
                      <td>k1111-7</td>
                      <td>k2222-7</td>
                      <td>k1111-7</td>
                    </tr>
                    <tr class="ux-table-row">
                      <td>k1111-8</td>
                      <td>k2222-8</td>
                      <td>k1111-8</td>
                      <td>k2222-8</td>
                      <td>k1111-8</td>
                    </tr>
                    <tr class="ux-table-row">
                      <td>k1111-9</td>
                      <td>k2222-9</td>
                      <td>k1111-9</td>
                      <td>k2222-9</td>
                      <td>k1111-9</td>
                    </tr>
                    <tr class="ux-table-row">
                      <td>k1111-10</td>
                      <td>k2222-10</td>
                      <td>k1111-10</td>
                      <td>k2222-10</td>
                      <td>k1111-10</td>
                    </tr>
                    <tr class="ux-table-row">
                      <td>k1111-11</td>
                      <td>k2222-11</td>
                      <td>k1111-11</td>
                      <td>k2222-11</td>
                      <td>k1111-11</td>
                    </tr>
                    <tr class="ux-table-row">
                      <td>k1111-12</td>
                      <td>k2222-12</td>
                      <td>k1111-12</td>
                      <td>k2222-12</td>
                      <td>k1111-12</td>
                    </tr>
                    <tr class="ux-table-row">
                      <td>k1111-13</td>
                      <td>k2222-13</td>
                      <td>k1111-13</td>
                      <td>k2222-13</td>
                      <td>k1111-13</td>
                    </tr>
                    <tr class="ux-table-row">
                      <td>k1111-14</td>
                      <td>k2222-14</td>
                      <td>k1111-14</td>
                      <td>k2222-14</td>
                      <td>k1111-14</td>
                    </tr>
                    <tr class="ux-table-row">
                      <td>k1111-15</td>
                      <td>k2222-15</td>
                      <td>k1111-15</td>
                      <td>k2222-15</td>
                      <td>k1111-15</td>
                    </tr>
                    <tr class="ux-table-row">
                      <td>k1111-16</td>
                      <td>k2222-16</td>
                      <td>k1111-16</td>
                      <td>k2222-16</td>
                      <td>k1111-16</td>
                    </tr>
                    <tr class="ux-table-row">
                      <td>k1111-17</td>
                      <td>k2222-17</td>
                      <td>k1111-17</td>
                      <td>k2222-17</td>
                      <td>k1111-17</td>
                    </tr>
                  </tbody>
                </table>
              </div>
            </div>
            <!---->
            <!---->
          </div>
        </div>
        <div class="ux-table-resize-line" style="display: none;"></div>
      </div>
    </div>
  </div>
</div>
`;

exports[`Table.virtual scroll scroll.y is number 1`] = `
<div class="ux-table-wrapper ux-table-light">
  <div class="ux-spin-nested-loading">
    <div class="ux-spin" style="display: none;"><svg viewBox="25 25 50 50" class="ux-spin-circle">
        <circle cx="50" cy="50" r="20" fill="none"></circle>
      </svg> </div>
    <div class="ux-spin-container">
      <div class="ux-table ux-table-fixed-header ux-table-scroll-position-left ux-table-default">
        <div class="ux-table-content">
          <div class="ux-table-main-table ux-table-scroll">
            <div class="ux-table-header">
              <table class="">
                <colgroup>
                  <col>
                  <col>
                  <col>
                  <col>
                  <col>
                </colgroup>
                <thead class="ux-table-thead">
                  <tr>
                    <th>title-0</th>
                    <th>title-1</th>
                    <th>title-2</th>
                    <th>title-3</th>
                    <th>title-4</th>
                  </tr>
                </thead>
                <!---->
              </table>
            </div>
            <div class="ux-table-body" style="max-height: 200px; overflow-y: scroll; webkit-transform: translate3d(0,0,0);">
              <div role="virtual" style="padding-top: 0px; padding-bottom: 7600px;">
                <table class="">
                  <colgroup>
                    <col>
                    <col>
                    <col>
                    <col>
                    <col>
                  </colgroup>
                  <tbody class="ux-table-tbody">
                    <tr class="ux-table-row">
                      <td>k1111-0</td>
                      <td>k2222-0</td>
                      <td>k1111-0</td>
                      <td>k2222-0</td>
                      <td>k1111-0</td>
                    </tr>
                    <tr class="ux-table-row">
                      <td>k1111-1</td>
                      <td>k2222-1</td>
                      <td>k1111-1</td>
                      <td>k2222-1</td>
                      <td>k1111-1</td>
                    </tr>
                    <tr class="ux-table-row">
                      <td>k1111-2</td>
                      <td>k2222-2</td>
                      <td>k1111-2</td>
                      <td>k2222-2</td>
                      <td>k1111-2</td>
                    </tr>
                    <tr class="ux-table-row">
                      <td>k1111-3</td>
                      <td>k2222-3</td>
                      <td>k1111-3</td>
                      <td>k2222-3</td>
                      <td>k1111-3</td>
                    </tr>
                    <tr class="ux-table-row">
                      <td>k1111-4</td>
                      <td>k2222-4</td>
                      <td>k1111-4</td>
                      <td>k2222-4</td>
                      <td>k1111-4</td>
                    </tr>
                    <tr class="ux-table-row">
                      <td>k1111-5</td>
                      <td>k2222-5</td>
                      <td>k1111-5</td>
                      <td>k2222-5</td>
                      <td>k1111-5</td>
                    </tr>
                    <tr class="ux-table-row">
                      <td>k1111-6</td>
                      <td>k2222-6</td>
                      <td>k1111-6</td>
                      <td>k2222-6</td>
                      <td>k1111-6</td>
                    </tr>
                    <tr class="ux-table-row">
                      <td>k1111-7</td>
                      <td>k2222-7</td>
                      <td>k1111-7</td>
                      <td>k2222-7</td>
                      <td>k1111-7</td>
                    </tr>
                    <tr class="ux-table-row">
                      <td>k1111-8</td>
                      <td>k2222-8</td>
                      <td>k1111-8</td>
                      <td>k2222-8</td>
                      <td>k1111-8</td>
                    </tr>
                    <tr class="ux-table-row">
                      <td>k1111-9</td>
                      <td>k2222-9</td>
                      <td>k1111-9</td>
                      <td>k2222-9</td>
                      <td>k1111-9</td>
                    </tr>
                  </tbody>
                </table>
              </div>
            </div>
            <!---->
            <!---->
          </div>
        </div>
        <div class="ux-table-resize-line" style="display: none;"></div>
      </div>
    </div>
  </div>
</div>
`;
